swarm/src/behaviour/peer_addresses.rs

```diff
@@ -0,0 +1,226 @@
+use crate::behaviour::FromSwarm;
+use crate::{DialError, DialFailure, NewExternalAddrOfPeer};
+
+use libp2p_core::Multiaddr;
+use libp2p_identity::PeerId;
+
+use lru::LruCache;
+
+use std::collections::HashSet;
+use std::num::NonZeroUsize;
+
+/// Struct for tracking peers' external addresses of the [`Swarm`](crate::Swarm).
+#[derive(Debug)]
+pub struct PeerAddresses(LruCache<PeerId, HashSet<Multiaddr>>);
+
+impl PeerAddresses {
+    /// Feed a [`FromSwarm`] event to this struct.
+    ///
+    /// Returns whether the event changed peer's known external addresses.
+    pub fn on_swarm_event(&mut self, event: &FromSwarm) -> bool {
+        match event {
+            FromSwarm::NewExternalAddrOfPeer(NewExternalAddrOfPeer { peer_id, addr }) => {
+                if let Some(peer_addrs) = self.0.get_mut(peer_id) {
+                    let addr = prepare_addr(peer_id, addr);
+                    peer_addrs.insert(addr)
+                } else {
+                    let addr = prepare_addr(peer_id, addr);
+                    self.put(*peer_id, std::iter::once(addr));
+                    true
+                }
+            }
+            FromSwarm::DialFailure(DialFailure {
+                peer_id: Some(peer_id),
+                error: DialError::NoAddresses,
+                ..
+            }) => self.0.pop(peer_id).is_some(),
+            _ => false,
+        }
+    }
+
+    pub fn new(cache_size: NonZeroUsize) -> Self {
+        Self(LruCache::new(cache_size))
+    }
+
+    pub fn put(&mut self, peer: PeerId, addresses: impl Iterator<Item = Multiaddr>) -> bool {
+        let addresses = addresses.filter_map(|a| a.with_p2p(peer).ok());
+        self.0.put(peer, HashSet::from_iter(addresses));
+        true
+    }
```

Comment:
This seems like a strange API to me, why would we overwrite all addresses that are already in there? I think we should replace this with an `add` function that simply adds the address to the cache.

Reply:
ach yes... my bad
:heavy_check_mark: